### PR TITLE
Adjunction v1 proxy

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by RyanGlScott** [comment]
Before discussing the finer details, let me first pose the question: why did you pair `V1` with `Proxy`? Given `V1`'s generics connotation, I would have expected `V1` to be paired with `U1`, if anything.

(I am aware that `U1` doesn't currently have a `Representable` instance, but I can try to make a new `distributive` release soon to make that a possibility.)

**Comment by phadej** [comment]
yeah, I'd rather picked `Void1` if one existed. `U1` would work equally well.

**Comment by RyanGlScott** [comment]
If we do choose `U1`, it might make sense to add instances for the other types in `GHC.Generics` too (e.g., `instance Adjunction Par1 Par1`). But this of course depends on #12 being merged to get them to have `Representable` instances.

**Comment by RyanGlScott** [comment]
Not that I'm aware of. I'd just define it locally like so:

```haskell
absurdV1 :: V1 a -> b
#if __GLASGOW_HASKELL__ >= 708
absurdV1 x = case x of {}
#else
absurdV1 !x = undefined
#endif
```

**Comment by ekmett** [comment]
I also prefer the `V1 -| U1` notion. It is very clean.

**Comment by phadej** [comment]
green

**Comment by RyanGlScott** [comment]
Looking good. But if we really want to complete the `GHC.Generics` collection, there's some more sensible instances you could add that mimic existing ones:

```haskell
instance Adjunction f g                     => Adjunction (Rec1 f) (Rec1 g)
instance Adjunction f g                     => Adjunction (M1 i c f) (M1 j d g)
instance (Adjunction f g, Adjunction f' g') => Adjunction (f' :.: f) (g :.: g')
instance (Adjunction f g, Adjunction f' g') => Adjunction (f :+: f') (g :*: g')
```

**Comment by ekmett** [comment]
Sadly, the M1 instances don't quite work due to uniqueness requirements.

**Comment by phadej** [comment]
@ekmett I have noticed that. Is `M1 i c f |- M1 i c g` good enough, or is it better to omit it?

**Comment by ekmett** [comment]
The semantics of it are wrong. c is meaningful content there.

**Comment by ekmett** [comment]
It describes things like constructor names, data type names, etc.

File: src/Data/Functor/Adjunction.hs [hunk 1]
```diff
@@ -194,3 +199,38 @@ instance Adjunction f u =>
...
+
+instance Adjunction Par1 Par1 where
+  leftAdjunct f = Par1 . f . Par1
+  rightAdjunct f = unPar1 . f . unPar1
+
+instance Adjunction f g => Adjunction (Rec1 f) (Rec1 g) where
+  unit   = Rec1 . leftAdjunct Rec1
+  counit = rightAdjunct unRec1 . unRec1
+
+-- @i@ and @c@ indexes have to be due functional dependency.
```
Comment (RyanGlScott) [inline src/Data/Functor/Adjunction.hs]: Perhaps a brief comment about why we don't define this instance?

**Comment by RyanGlScott** [comment]
Thanks, @phadej!

**Pull request merged**